src/hotspot/share/opto/escape.cpp

```diff
@@ -387,6 +398,318 @@ bool ConnectionGraph::compute_escape() {
+
+          if (child->is_Store()) {
+            assert(child->in(0) != NULL || m->in(0) != NULL, "No control for store or AddP.");
+            if (_igvn->is_dominator(merge_phi_region, child->in(0) != NULL ? child->in(0) : m->in(0))) {
+              return false;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+bool ConnectionGraph::can_reduce_this_phi(const Node* phi) const {
+  if (!is_ideal_node_in_graph(phi->_idx)) return false;
+  if (ptnode_adr(phi->_idx)->escape_state() != PointsToNode::EscapeState::NoEscape) return false;
+
+  const Type* phi_t  = _igvn->type(phi);
+
+  // Found a Memory edge coming from the same Region as the Phi
+  bool found_memory_edge = false;
+
+  // Is any of the users of the Phi a Call node?
+  bool has_call_as_user = false;
+
+  // Ignoring any ConP#Null, is there any of the Phi inputs Non Scalar Replaceable?
+  bool has_nonnull_nonsr_input = false;
+
+  // Ignoring ConP#Null inputs, are all the inputs to the Phi of the same Klass?
+  bool mixed_klasses = false;
+
+
+  // If not an InstPtr bail out
+  if (phi_t == NULL || phi_t->make_oopptr() == NULL || phi_t->make_oopptr()->isa_instptr() == NULL) {
+    return false;
+  }
+
+  // Validate inputs:
+  //    Check whether this Phi node actually point to any scalar replaceable
+  //    Allocate node of the same Klass as the Phi.
+  //    Also checks that there is no write to any of the inputs after the
+  //    merge occurs.
+  bool has_noescape_allocate = false;
+  ciInstanceKlass* klass = phi_t->make_oopptr()->is_instptr()->instance_klass();
+  for (uint in_idx = 1; in_idx < phi->req(); in_idx++) {
+    // come_from_allocate returns NULL if the source isn't an Allocate
+    const Node* input = come_from_allocate(phi->in(in_idx));
+    PointsToNode* input_ptn = input != NULL ? ptnode_adr(input->_idx) : NULL;
+
+    // Check if input comes from scalar replaceable Allocate
+    bool is_sr_input = (input_ptn != NULL && input_ptn->scalar_replaceable());
+    has_nonnull_nonsr_input |= !is_sr_input;
+    has_noescape_allocate |= is_sr_input;
+
+    // Check if there is no write to the input after it is merged.
+    // If there is a write to any input after the merge we need to bail out.
+    if (!is_read_only(phi->in(0), phi->in(in_idx))) {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. The %dth input has a store after the merge.", phi->_idx, in_idx);)
+      return false;
+    }
+
+    const Type* input_t = _igvn->type(phi->in(in_idx))->make_oopptr();
+    if (input_t == NULL || input_t->isa_instptr() == NULL) {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. The %dth input is not an InstPtr.", phi->_idx, in_idx);)
+      return false;
+    }
+
+    if (klass != input_t->is_instptr()->instance_klass()) {
+      mixed_klasses = true;
+    }
+
+    klass = input_t->is_instptr()->instance_klass();
```

Comment:
Why you are updating `Klass` value here?

Reply:
My mistake. No need for that.